certbot/interfaces.py

```diff
@@ -604,3 +604,204 @@ def renew_deploy(self, lineage, *args, **kwargs):
+    """
+
+    @abc.abstractmethod
+    def arguments(self):
+        """
+        Returns a list of arguments for this ParserNode object
+
+        :returns: list of arguments
+        """
+
+    @abc.abstractmethod
+    def dirty(self):
+        """
+        Returns a boolean telling if the state of the node was changed after the
+        initial read or the last save operation. The main usage for this method is
+        to discover the nodes in the tree that need to be written to disk upon
+        save.
+
+        :param str name: The name of the child node to add
+
+        :returns: True if the node needs to be written to disk
+        """
+
+    @abc.abstractmethod
+    def save(self, msg):
+        """
+        Save traverses the children, and attempts to write the AST to disk for
+        all the objects that are marked dirty. The actual operation of course
+        depends on the underlying implementation. save() shouldn't be called
+        from the Configurator outside of it's designated save() method in order
+        to ensure that the Reverter checkpoints are created properly.
+
+        Note: this approach of keeping internal structure of the configuration
+        within the ParserNode tree does not represent the file inclusion structure
+        of actual configuration files that reside in the filesystem. To handle
+        file writes properly, the file specific temporary trees should be extracted
+        from the full ParserNode tree where necessary when writing to disk.
+
+        """
+
+    @abc.abstractmethod
+    def set_arguments(self, arguments):
+        """
+        Sets argument list of the ParserNode object and marks the node dirty.
+        """
+
+
+@six.add_metaclass(abc.ABCMeta)
+class CommentNode(ParserNode):
+    """
+    CommentNode class is used for representation of comments within the parsed
+    configuration structure. Because of the nature of comments, it is not able
+    to have child nodes and hance it is always treated as a leaf node.
+    """
+
+
+@six.add_metaclass(abc.ABCMeta)
+class DirectiveNode(ParserNode):
+    """
+    DirectiveNode class represents a configuration directive within the configuration.
+    It can have zero or more arguments attached to it. Because of the nature of
+    single directives, it is not able to have child nodes and hence it is always
+    treated as a leaf node.
+    """
+
+
+@six.add_metaclass(abc.ABCMeta)
+class BlockNode(ParserNode):
+    """
+    BlockNode class represents a block of nested configuration directives, comments
+    and other blocks as its children. A BlockNode can have zero or more arguments
+    attached to it.
+
+    Configuration blocks typically consist of one or more child nodes of all possible
+    types. Because of this, the BlockNode class has various discovery and structure
+    management methods.
+    """
+
+    @abc.abstractmethod
+    def add_child_block(self, name, arguments=[], position=-1):
```

Comment:
This is definitely a nitpick, but I don't love how the position parameter deviates from `list.insert` in the standard library. As an example:
```
In [1]: l = [0,1,2]                                                                                                                                                                                                                                                                                                                                                                                                                   

In [2]: l.insert(-1, 3)                                                                                                                                                                                                                                                                                                                                                                                                               

In [3]: l                                                                                                                                                                                                                                                                                                                                                                                                                             
Out[3]: [0, 1, 3, 2]
```
What if we keep the behavior the same but make the default value of `position` `None` which appends to the end of the list?

Reply:
Also to satisfy our linters concerns (which I think are valid), we should make the default value of `arguments` `None` as well.

Reply:
Using `None` as the default value for both makes sense, these are changed now.